### PR TITLE
feat(security): Create a Vault mgmt token for Consul Secrets API Operations

### PR DESCRIPTION
**Pull request by jim-wang-intel**
Part of Phase 1 Securing Consul stories: Create a Vault token to configure consul access

Changes in `secretstore-setup`:
 - add token file writer implementation for creating and writing Vault's Consul secrets admin token
 - add Consul ACL feature flag `"ENABLE_REGISTRY_ACL"`
 - add configuration setting for `ConsulSecretAdminTokenPath` for specifying token file location
 - refactor TokenMaintenance to share code with token file writer's implementation

Closes: #3155

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/master/.github/Contributing.md.

## What is the current behavior?

N/A

## What is the new behavior?
New Vault token created for special purpose of management Consul Secrets APIs operations policy: configure access / create update roles later on.  This token will be later used by Consul's bootstrapper.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x ] No

## New Imports

- [ ] Yes
- [x ] No

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?

## Other information
To verify this locally, one can git clone this PR and build a dev version of `docker_secretstore_setup` and then add the following environment variables to docker-compose file of `secretstore-setup`:
```yaml
 environment:
      ENABLE_REGISTRY_ACL: "true"
```
so that the ACL feature is turned on.  Run `make run dev` and then observe the following docker logs from `edgex-secretstore-setup`:
`docker logs edgex-secretstore-setup | grep "edgex-consul/admin/token"`

```

jim@jim-NUC7i5DNHE:~/go/src/github.com/edgexfoundry/edgex-compose/compose-builder$ docker logs edgex-secretstore-setup | grep "edgex-consul/admin/token"
level=INFO ts=2021-02-24T22:47:57.466345924Z app=edgex-security-secretstore-setup source=tokenfilewriter.go:110 msg="token's written to /tmp/edgex/secrets/edgex-consul/admin/token.json"
jim@jim-NUC7i5DNHE:~/go/src/github.com/edgexfoundry/edgex-compose/compose-builder$ 

```
And also list the file, too:
`$ sudo ls -al /tmp/edgex/secrets/edgex-consul/admin/`
```
[sudo] password for jim: 
total 12
drwx------ 2 2002 2001 4096 Feb 24 12:56 .
drwx------ 3 2002 2001 4096 Feb 24 12:56 ..
-rw------- 1 2002 2001  536 Feb 24 15:47 token.json
```

### CODE REVIEW

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	if fileErr = json.NewEncoder(fileWriter).Encode(tokenCreated); fileErr != nil {
+		_ = fileWriter.Close()
+		return nil, fmt.Errorf("failed to write created token: %s", fileErr.Error())
+	}
+
+	if fileErr = fileWriter.Close(); fileErr != nil {
+		return nil, fmt.Errorf("failed to close token file: %s", fileErr.Error())
+	}
+
+	w.logClient.Infof("token's written to %s", tokenFilePath)
```
Comment (lenny-intel) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:110]: A nit, but should `token's` be `token` since it is writing a single token?

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+	return revokeTokenFunc, nil
+}
+
+// CreateMgmtTokenForConsulSecretsEngine creates a new Vault token that
+// allows the Consul bootstrapper to operate on managing Vault's Consul secrets engine related APIs (see reference:
+// https://www.vaultproject.io/api-docs/secret/consul). The created Vault token is meant for serving
+// the purpose of Consul ACL's bootstrapping as part of securing Consul process.
+//
+// Requires a root token to create, and returns data/information containing the token,
+// keeping the token without revoking it and hence always returning nil RevokeFunc;
```
Comment (lenny-intel) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:121]: Why have `RevokeFunc`  when always returning nil for it? Is it due to function signature? If so add this `why` to this description.

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+			}
+		})
+	}
+}
+
+func TestFileWriteErrorForConsulSecretEngine(t *testing.T) {
+	curUser, _ := user.Current()
+	if curUser != nil && curUser.Uid == "0" {
+		// it is root user then we skip this test as root can have permission to write everything
+		t.Skip()
```
Comment (lenny-intel) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:149]: Should we fail rather than skip so developer is forced to not run the tests as root?

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	// as only the root user can perform this
+	testTokenFileDir := "/root/test"
+	testTokenFilePath := filepath.Join(testTokenFileDir, "mgmt-token.json")
+	fileWriter := NewWriter(lc, secretClient, flOpener)
+	_, err := fileWriter.CreateAndWrite(testRootToken, testTokenFilePath,
+		fileWriter.CreateMgmtTokenForConsulSecretsEngine)
+	require.Error(t, err)
+}
+
+func setup() func() (logger.LoggingClient, fileioperformer.FileIoPerformer) {
```
Comment (lenny-intel) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:176]: This should be TestMain() which is the standard for this type of setup and doesn't requires a call in each test. `lc` and `flOpener`  become package level globals.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 1 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:110]: yea, i meant it to be `token is written` i'll change it to spell it out

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 2 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:121]: yes, it is due to function signature. I'll add the comments

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go (hunk 3 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:149]: not sure, if CI is running as root then the test will be failing. Is CI running as root user?

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go (hunk 4 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:176]: ok

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go (hunk 3 above)
Comment (lenny-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:149]: I doubt it, but @ernestojeda should be able to answer the question.

Comment (ernestojeda) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:149]: Yes, when `make test` runs, it runs inside the context of a docker image and we run as root due to permissions issues with the LF build agents. See: https://github.com/edgexfoundry/edgex-global-pipelines/blob/master/vars/edgeXBuildGoApp.groovy#L181

Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:149]: Thank you, @ernestojeda.  @lenny-intel i'll see this comment as resolved then.

File: cmd/security-secretstore-setup/res/configuration.toml [hunk 5]
```diff
@@ -38,6 +38,7 @@ TokenProviderAdminTokenPath = "/run/edgex/secrets/tokenprovider/secrets-token.js
 PasswordProvider = ""
 PasswordProviderArgs = [ ]
 RevokeRootTokens = true
+ConsulSecretsAdminTokenPath = "/tmp/edgex/secrets/edgex-consul/admin/token.json"
```
Comment (bnevis-i) [inline cmd/security-secretstore-setup/res/configuration.toml:41]: Don't code the whole path into the executable.  Use existing environment variables.

File: internal/security/secretstore/constants.go [hunk 6]
```diff
@@ -19,10 +19,8 @@
 package secretstore
 
 const (
-	// KVSecretsEngineMountPoint is the name of the mount point base for Vault's key-value secrets engine
-	KVSecretsEngineMountPoint = "secret"
-	// ConsulSecretEngineMountPoint is the name of the mount point base for Vault's Consul secrets engine
-	ConsulSecretEngineMountPoint = "consul"
+	// this is a feature key to indicate whether to enable Registry Consul's ACL or not
+	RegistryACLFeatureFlag = "ENABLE_REGISTRY_ACL"
```
Comment (bnevis-i) [inline internal/security/secretstore/constants.go:23]: Remove. This feature doesn't need to be in secretstore setup.  Secretstore-setup doesn't enable the Consul ACL feature.  We can create the token for configuring the consul secrets provider regardless of whether or not we enable Consul ACLS.

File: internal/security/secretstore/init.go [hunk 7]
```diff
@@ -422,6 +422,19 @@ func (b *Bootstrap) BootstrapHandler(ctx context.Context, _ *sync.WaitGroup, _ s
 		lc.Info("proxy certificate pair upload was skipped because cert secretStore value(s) were blank")
 	}
 
+	// If Registry Consul ACL feature is enabled, then we need to create and save a Vault token to configure
+	// Consul secret engine access, role operations, and managing Consul agent tokens.
+	registryACLEnabled := os.Getenv(RegistryACLFeatureFlag)
+	if registryACLEnabled == "true" {
```
Comment (bnevis-i) [inline internal/security/secretstore/init.go:428]: Do this every time.  It is harmless.

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+	if fileErr != nil {
+		return nil, fmt.Errorf("failed to create tokenpath base dir: %s", fileErr.Error())
+	}
+
+	fileWriter, fileErr := w.fileOpener.OpenFileWriter(tokenFileAbsPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600)
+	if fileErr != nil {
+		return nil, fmt.Errorf("failed to create token file %s: %s", tokenFileAbsPath, fileErr.Error())
+	}
+
+	if fileErr = json.NewEncoder(fileWriter).Encode(tokenCreated); fileErr != nil {
```
Comment (bnevis-i) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:101]: This line can fail if NewEncoder() returns nil.

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+		_ = fileWriter.Close()
+		return nil, fmt.Errorf("failed to write created token: %s", fileErr.Error())
+	}
+
+	if fileErr = fileWriter.Close(); fileErr != nil {
```
Comment (bnevis-i) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:106]: See internal/security/fileprovider/provider.go lines 172-196.  Since we are passing the token to another container, we want to be able to configure file permissions, uid, and gid.

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 2 above)
Comment (bnevis-i) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:121]: For creating the token creating token used by filetokenprovider, we revoke to token after it is used via the `RevokeFunc`.  I needs to be here for the other use case.

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+	if err := w.secretClient.InstallPolicy(rootToken,
+		consulSecretsEngineOpsPolicyName,
+		consulSecretsEngineOpsPolicyDocument); err != nil {
+		return nil, nil, fmt.Errorf("failed to install Consul secrets engine operations policy: %v", err)
+	}
+
+	// setup new token's properties
+	tokenParams := make(map[string]interface{})
+	tokenParams["type"] = "service"
+	tokenParams["display_name"] = "managing with policy name " + consulSecretsEngineOpsPolicyName
```
Comment (bnevis-i) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:149]: Suggest display_name = "Configuration token for consul secrets engine"

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+			expectError := test.installPolicyCallErr != nil ||
+				test.createTokenCallErr != nil ||
+				len(test.rootToken) == 0
+
+			if expectError {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				require.FileExists(t, testTokenFilePath)
```
Comment (bnevis-i) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:139]: I don't understand.  If we're using a mockable FileOpener, why is an actual file created?

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+	fileWriter := NewWriter(lc, secretClient, flOpener)
+	_, err := fileWriter.CreateAndWrite(testRootToken, testTokenFilePath,
+		fileWriter.CreateMgmtTokenForConsulSecretsEngine)
+	require.Error(t, err)
+}
+
+func setup() func() (logger.LoggingClient, fileioperformer.FileIoPerformer) {
+	return func() (logger.LoggingClient, fileioperformer.FileIoPerformer) {
+		lc := logger.MockLogger{}
+		flOpener := fileioperformer.NewDefaultFileIoPerformer()
```
Comment (bnevis-i) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:179]: Can you pass a mock object here instead of NewDefaultFileIoPerformer so that you can test file reading and writing without actually having to read and write files?

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go (hunk 3 above)
Comment (bnevis-i) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:149]: See my other comment about mocking the file creation commands and thus having to avoid this debacle althogether.

**Review by bnevis-i — CHANGES_REQUESTED** [review]

File: cmd/security-secretstore-setup/res/configuration.toml (hunk 5 above)
Comment (jim-wang-intel) [reply cmd/security-secretstore-setup/res/configuration.toml:41]: don't get it, what existing env?

File: internal/security/secretstore/constants.go (hunk 6 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/constants.go:23]: I disagree, if no ACL enable why need to create this token at all?

File: internal/security/secretstore/init.go (hunk 7 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/init.go:428]: same comment above, no need to create useless token if no ACL

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 8 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:101]: could not think of any case, could you give me a concrete example that json.NewEncoder will fail?

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 10 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:149]: i think our go.mod.secret code in Vault's CreateToken() automatically prefaces `token-`  -xxxxxx something or Vault itself does it, so with my example above, the display_name will be `token-managing-with-policy-name-xxxxxxx`,  that is why I keep the display name this way.

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go (hunk 11 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:139]: we are not using mockable fileopener here, concrete file creates

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go (hunk 12 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:179]: i don't see a need here as there are already tests in`NewDefaultFileIoPerformer`  itself doing the mocking test why duplicate tests here?

File: cmd/security-secretstore-setup/res/configuration.toml (hunk 5 above)
Comment (bnevis-i) [reply cmd/security-secretstore-setup/res/configuration.toml:41]: Sorry, it wasn't an env var.  It was OutputDir, here https://github.com/edgexfoundry/edgex-go/blob/12a188d7528ee379d18cf3e1ac715e90dfef7258/cmd/security-secretstore-setup/res-file-token-provider/configuration.toml#L11

Comment (jim-wang-intel) [reply cmd/security-secretstore-setup/res/configuration.toml:41]: ok

Comment (jim-wang-intel) [reply cmd/security-secretstore-setup/res/configuration.toml:41]: but that is in fileprovider, nothing to with that utility here

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 9 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:106]: I don't understand this: the existing code for creating token-issuing token doesn't have this logic at all;  are you suggesting we modify the existing privilege for token issuing token , too???

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 8 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:101]: @bnevis-i :   It is impossible for `json.NewEncoder` to return `nil`, please check out the golang source code at https://golang.org/src/encoding/json/stream.go  line 191-194, in which it is always return an instance of `&Encoder{w: w, escapeHTML: true}`.  Thus `json.NewEncoder()` will never return nil.

Comment (bnevis-i) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:101]: Good to know.

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 1 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go]: fixed

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 2 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go]: added comments

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go (hunk 4 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:177]: added TestMain()

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 9 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:106]: @bnevis-i  need more discussion on this. I do not quite understand the logic in the line 172-196 you mentioned above there and I thought the file permission already specified by the fileOpenWriter.  We are not using this token to create another token so I don't understand why we need to do that step. Let's have a discussion tomorrow morning if you have time.

Comment (bnevis-i) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:106]: fileOpenWriter only specifies the mode.  If we are running containers as non-root users the file owner and group must match that of the consuming container.

Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:106]: Hi @bnevis-i  if the `secretstore-setup` is running as non-root, then the chown operations won't work, right?  Also as of today, this code in the entrypoint script running as root, https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-secretstore-setup/entrypoint.sh#L33-L36  would change the ownership to 2002:2001 anyway, so that the ownership change in the provider is a moot point.

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go [hunk 13]
```diff
@@ -32,17 +32,21 @@ import (
 	"github.com/edgexfoundry/go-mod-secrets/v2/secrets/mocks"
 )
 
-func TestNewTokenFileWriter(t *testing.T) {
-	lc, flOpener := setup()()
+var lc logger.LoggingClient
+var flOpener fileioperformer.FileIoPerformer
+
+func TestMain(m *testing.M) {
+	lc, flOpener = setup()()
```
Comment (lenny-intel) [inline internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:39]: Why not move code from setup() into here? It is just a couple line that can be here.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: cmd/security-secretstore-setup/res/configuration.toml (hunk 5 above)
Comment (jim-wang-intel) [reply cmd/security-secretstore-setup/res/configuration.toml:41]: resolved

File: internal/security/secretstore/constants.go (hunk 6 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/constants.go:23]: resolved, also use this same flag to not install Cosnul's secret engine

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 9 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:106]: resolve as we discussed

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter.go (hunk 10 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter.go:150]: "token" + "display_name":  **for Consul ACL bootstrap**
resolved

File: internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go (hunk 13 above)
Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go:39]: ok, i'll move it.

Comment (jim-wang-intel) [reply internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go]: fixed

**Review by lenny-intel — APPROVED** [review]
LGTM
